src/Symfony/Bundle/FrameworkBundle/Routing/DelegatingLoader.php

```diff
@@ -73,14 +73,36 @@ public function load($resource, $type = null)
         }
 
         foreach ($collection->all() as $route) {
-            if (!is_string($controller = $route->getDefault('_controller')) || !$controller) {
+            if (!is_string($controller = $route->getDefault('_controller'))) {
                 continue;
             }
 
-            try {
-                $controller = $this->parser->parse($controller);
-            } catch (\InvalidArgumentException $e) {
-                // unable to optimize unknown notation
+            if (false !== strpos($controller, '::')) {
+                continue;
+            }
+
+            if (2 === substr_count($controller, ':')) {
+                $deprecatedNotation = $controller;
+
+                try {
+                    $controller = $this->parser->parse($controller);
+
+                    @trigger_error(sprintf(
+                        'Referencing controllers with %s is deprecated since version 4.1 and will be removed in 5.0. Use %s instead.',
+                        $deprecatedNotation,
+                        $controller
+                    ), E_USER_DEPRECATED);
+                } catch (\InvalidArgumentException $e) {
+                    // unable to optimize unknown notation
+                }
+            }
+
+            if (1 === substr_count($controller, ':')) {
```

Comment:
You could use an `elseif` here if the condition is mutually exclusive with the one above. It would prevent evaluating this one when the first is true.

Reply:
Depends if someone used an extended name parser that returned something else. I would leave it like this for simplicitly.